File: src/Week1/pset1/credit.c
#include <cs50.h>
#include <stdio.h>

const int american_express_card = 1;
const int mastercard_card = 2;
const int visa_card = 3;

int main(void)
{
    long card_number = get_long("Number: ");
    long temp = card_number;
    int count = 0;
    int last_digit_series_sum = 0;
    int second_last_digit_series_sum = 0;

    //determines of which company the card will be
    int card_type = 0;

    //counts number of digits in the card
    while (temp > 0)
    {
        count++;
        temp /= 10;
    }

    if (count != 13 && count != 15 && count != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    if (count == 15)
    {
        //10^13
        int first_two_digits = card_number / 10000000000000;
        if (first_two_digits == 34 || first_two_digits == 37)
        {
            card_type = 1;
        }
    }
    else if (count == 13)
    {
        //10^12
        int first_digit = card_number / 1000000000000;
        if (first_digit == 4)
        {
            card_type = 3;
        }
    }
    else
    {
        //10^15
        int first_digit = card_number / 1000000000000000;
        if (first_digit == 4)
        {
            card_type = 3;
        }
        else
        {
            //10^14
            int first_two_digits = card_number / 100000000000000;
            if (first_two_digits == 51 || first_two_digits == 52 || first_two_digits == 53
                || first_two_digits == 54 || first_two_digits == 55)
            {
                card_type = 2;
            }
        }
    }
    //Don't validate checksum if the card_type is still none
    if (card_type == 0)
    {
        printf("INVALID\n");
        return 0;
    }
    //validates for checksum
    while (card_number > 0)
    {
        //adding which are at odd digits place
        last_digit_series_sum += card_number % 10;

        //removing the processed digit to access the numbers at even position starting from second last digit
        card_number /= 10;

        int remainder = (card_number % 10) * 2;

        if (remainder > 9)
        {
            second_last_digit_series_sum += remainder / 10 + remainder % 10;
        }
        else
        {
            second_last_digit_series_sum += remainder;
        }

        //removing the processed digit at even's position
        card_number /= 10;
    }

    if ((last_digit_series_sum + second_last_digit_series_sum) % 10 == 0)
    {
        if (card_type == american_express_card)
        {
            printf("AMEX\n");
        }
        else if (card_type == mastercard_card)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("VISA\n");
        }
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: src/Week6/pset6/credit/credit.py
from cs50 import get_string
from sys import exit
from math import floor

# AMERICAN EXPRESS
AE = ["34", "37"]

# MASTERCARD
MC = ["51", "52", "53", "54", "55"]

# VISA
VISA = "4"

brand = "INVALID"
card = get_string("Number: ")

# remove any trailing spaces
card = card.strip()
card_length = len(card)

if card_length not in [13, 15, 16]:
    print(brand)
    exit(0)

if card_length == 15:
    if card[:2] in AE:
        brand = "AMEX"
elif card_length == 13:
    if card[:1] == VISA:
        brand = "VISA"
else:
    if card[:2] in MC:
        brand = "MASTERCARD"
    elif card[:1] == VISA:
        brand = "VISA"

# Don't validate checksum if brand still remains INVALID
if brand == "INVALID":
    print(brand)
    exit(0)

# casting string to int
card = int(card)
second_last_digits_series_sum = 0
last_digits_series_sum = 0
while card > 0:
    last_digits_series_sum += card % 10
    # removing the already processed digit
    card /= 10
    # remove the decimal point and it's post digit
    card = floor(card)
    remainder = (card % 10) * 2

    if remainder > 9:
        second_last_digits_series_sum += remainder % 10
        remainder = floor(remainder / 10)
        second_last_digits_series_sum += remainder
    else:
        second_last_digits_series_sum += remainder

    card /= 10
    card = floor(card)

if (last_digits_series_sum + second_last_digits_series_sum) % 10 == 0:
    print(brand)
else:
    print("INVALID")